nitransforms/surface.py

```diff
@@ -0,0 +1,124 @@
+
+    def __init__(self, mat):
+        """Initialize the transform.
+
+        Parameters
+        ----------
+        mat : array-like, shape (nv1, nv2)
+            Sparse matrix representing the transform.
+        """
+        super().__init__()
+        if isinstance(mat, sparse.csr_array):
+            self.mat = mat
+        else:
+            self.mat = sparse.csr_array(mat)
+
+    def apply(self, x, inverse=False, normalize="element"):
+        """Apply the transform to surface data.
+
+        Parameters
+        ----------
+        x : array-like, shape (..., nv1)
+            Data to transform.
+        inverse : bool, default=False
+            Whether to apply the inverse transform. If True, ``x`` has shape
+            (..., nv2), and the output will have shape (..., nv1).
+        normalize : {"element", "sum", "none"}, default="element"
+            Normalization strategy. If "element", the scale of each value in
+            the output is comparable to each value of the input. If "sum", the
+            sum of the output is comparable to the sum of the input. If
+            "none", no normalization is applied.
+
+        Returns
+        -------
+        y : array-like, shape (..., nv2)
+            Transformed data.
+        """
+        if normalize not in ("element", "sum", "none"):
+            raise ValueError("Invalid normalization strategy.")
+
+        mat = self.mat.T if inverse else self.mat
+
+        if normalize == "element":
+            sum_ = mat.sum(axis=0)
+            scale = np.zeros_like(sum_)
+            mask = sum_ != 0
+            scale[mask] = 1.0 / sum_[mask]
+            mat = mat @ sparse.diags(scale)
+        elif normalize == "sum":
+            sum_ = mat.sum(axis=1)
+            scale = np.zeros_like(sum_)
+            mask = sum_ != 0
+            scale[mask] = 1.0 / sum_[mask]
+            mat = sparse.diags(scale) @ mat
+
+        y = x @ mat
+        return y
+
+    def _to_hdf5(self, x5_root):
+        """Write transform to HDF5 file."""
+        xform = x5_root.create_group("Transform")
+        xform.attrs["Type"] = "surface"
+        xform.create_dataset("data", data=self.mat.data)
+        xform.create_dataset("indices", data=self.mat.indices)
+        xform.create_dataset("indptr", data=self.mat.indptr)
+        xform.create_dataset("shape", data=self.mat.shape)
+
+    def to_filename(self, filename, fmt=None):
+        """Store the transform."""
+        if fmt is None:
+            fmt = "npz" if filename.endswith(".npz") else "X5"
+
+        if fmt == "npz":
+            sparse.save_npz(filename, self.mat)
+            return filename
+
+        with h5py.File(filename, "w") as out_file:
+            out_file.attrs["Format"] = "X5"
+            out_file.attrs["Version"] = np.uint16(1)
+            root = out_file.create_group("/0")
+            self._to_hdf5(root)
+
+        return filename
+
+    @classmethod
+    def from_filename(cls, filename, fmt=None):
```

Comment:
from filename would not read sphere.reg and so forth?

Reply:
This one is only reading the sparse matrix from a file.
With the code Dylan's working on, we can generate the sparse matrix from the spheres.